hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -1009,6 +1013,66 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 	return ctrl.Result{}, nil
 }
 
+func (r *NodePoolReconciler) cleanupMachineTemplates(ctx context.Context, log logr.Logger, nodePool *hyperv1.NodePool, controlPlaneNamespace string) error {
+	// list machineSets
+	machineSets := &capiv1.MachineSetList{}
+	if err := r.Client.List(ctx, machineSets, client.InNamespace(controlPlaneNamespace)); err != nil {
+		return fmt.Errorf("failed to list machineSets: %w", err)
+	}
+
+	// filter machineSets owned by this nodePool.
+	nodePoolKey := client.ObjectKeyFromObject(nodePool).String()
+	filtered := make([]*capiv1.MachineSet, 0, len(machineSets.Items))
+	for idx := range machineSets.Items {
+		ms := &machineSets.Items[idx]
+		// skip if machineSet doesn't belong to the nodePool
+		if ms.Annotations[nodePoolAnnotation] != nodePoolKey {
+			continue
+		}
+
+		filtered = append(filtered, ms)
+	}
+
+	if len(filtered) == 0 {
+		// initial machineSet has not been created.
+		log.Info("initial machineSet has not been created.")
+		return nil
+	}
+
+	ref := filtered[0].Spec.Template.Spec.InfrastructureRef
+	machineTemplates := new(unstructured.UnstructuredList)
+	machineTemplates.SetAPIVersion(ref.APIVersion)
+	machineTemplates.SetKind(ref.Kind)
+	if err := r.Client.List(ctx, machineTemplates, client.InNamespace(ref.Namespace)); err != nil {
+		return fmt.Errorf("failed to list MachineTemplates: %w", err)
+	}
+
+	// delete old machine templates not currently referenced by any machineSet.
+	for _, mt := range machineTemplates.Items {
+		// skip if MachineTempalte doesn't belong to the nodePool.
```

Comment:
nit: to avoid n^2 time complexity we could run two separate loops, one that stores the templates for the NodePool, the second one stores the refs from the existing machineSets. No need to change.